### DISCUSSION TITLE
What are all the prebuilt formats possible

### DISCUSSION THREAD

**Discussion by Overloader6** [nasa/openmct#4135] [Community Q&A] [answered]
I couldnt find anywhere in the documentation what formats there are you can use. I know of float, utc and enum, i guess there is probably a string (will try after this) but for example is there a format for bool values?

**Comment by jvigliotta** [comment]
HI @Voreck could you provide a little more detail? What are you trying to accomplish?

**Reply by Overloader6** [reply]
Honestly not trying to accomplish anything with that. I just know that there are some prebuilt options like float, enum, utc and string and that you can add your own formats. So I just would like to know what formats there already are besides the one i listed

**Reply by jvigliotta** [reply]
Sorry @Voreck for the very late reply. We don't have a set list of built-in formats. The API has some information here: https://github.com/nasa/openmct/blob/master/API.md#registering-formats for registering custom formats, which may be helpful if you haven't seen it yet.

**Reply by inad9300** [reply]
I'm also interested in the list of built-in formats. Also in knowing what the default formatter does.

I would additionally pose the question: how come documentation is so poor in a project like this, to the point where even maintainers can't find basic information? Lots to improve in this regard, I reckon.

**Reply by ozyx** [reply, marked as answer]
> I'm also interested in the list of built-in formats. Also in knowing what the default formatter does.

Hi @VoreckLukas and @inad9300 ,

Thank you for bringing this to our attention. We genuinely appreciate feedback from our community and recognize that our documentation can certainly be improved. To answer your queries:

The core of a formatter in Open MCT is defined by the `Formatter` interface (I've provided TypeScript-style interfaces for simplicity):

```ts
interface Formatter {
  key: string;
  parse: (value: any) => any;
  format: (value: any) => string;
  validate: (value: any) => boolean;
}
```

Open MCT on its own defines three built-in formatters:
1. **[Number (default)](https://github.com/nasa/openmct/blob/bd7ad531955d98bf920712090db41f6290e13eab/src/api/telemetry/TelemetryValueFormatter.js#L26):** 

Applied to data with `format: 'number'`
```js
valueMetadata = {
    format: 'number'
    // ...
};
```

```ts
interface NumberFormatter extends Formatter {
    parse: (x: any) => number;
    format: (x: number) => string;
    validate: (value: any) => boolean;
}
```
2. **[String](https://github.com/nasa/openmct/blob/bd7ad531955d98bf920712090db41f6290e13eab/src/api/telemetry/TelemetryValueFormatter.js#L96)**:

Applied to data with `format: 'string'`
```js
valueMetadata = {
    format: 'string'
    // ...
};
```
```ts
interface StringFormatter extends Formatter {
    parse: (value: any) => string;
    format: (value: string) => string;
    validate: (value: any) => boolean;
}
```

3. **[Enum](https://github.com/nasa/openmct/blob/bd7ad531955d98bf920712090db41f6290e13eab/src/api/telemetry/TelemetryValueFormatter.js#L55)**:
Applied to data with `format: 'enum'`
```js
valueMetadata = {
    format: 'enum',
    enumerations: [
    {
        value: 1,
        string: 'APPLE'
    }, 
    {
        value: 2,
        string: 'PEAR',
    },
    {
        value: 3,
        string: 'ORANGE'
    }]
    // ...
};
```

Creates a two-way mapping between enum string and value to be used in the `parse` and `format` methods.
Ex:
- `formatter.parse('APPLE') === 1;`
- `formatter.format(1) === 'APPLE';`

```ts
interface EnumFormatter extends Formatter {
    parse: (value: string) => string;
    format: (value: number) => string;
    validate: (value: any) => boolean;
}
```

Regarding your concerns about our documentation, we understand the frustration. With a project as long-standing as Open MCT, there are numerous areas of documentation that require updating, and your feedback serves as a valuable reminder. I will be updating our Formatter documentation immediately with the information I've provided here.

Hope this helps! Please feel free to reach out with any other questions. Thanks for being a part of the Open MCT community!

**Reply by inad9300** [reply]
Thanks so much for the information and the proactiveness.

I noticed that, in your answer, you wrote a three-element list containing elements 1, 4 and 5. Why the gap? Perhaps related, there are examples in the API documentation and other places using formats "utc", "float" and "integer".

**Reply by ozyx** [reply]
Whoops! That was a typo. Nice catch.

The `utc` format is provided by the [UTCTimeSystem plugin](https://github.com/nasa/openmct/blob/2940fb471e56ec91f54ec972cbf3836652f4ec91/src/plugins/utcTimeSystem/UTCTimeFormat.js#L33), and docs can be found inline.

As for `'integer'` and `'float'`, those may have just been used as examples in the documentation (this particular area of which was penned a little over 6 years ago) as I don't find them defined anywhere.
